README.md
# QBOX

## Description

A Bulletin Board System for the Sinclair QL, originally written in 1987-94 for analogue modems. But thanks to its **QSPIL** interface, it can be made to work with TCP/IP too! You just need an emulator supporting the TCP device. QBOX has been tested successfully with [QPC2](https://www.kilgus.net/qpc/), [SMSQmulator](http://www.wlenerz.com/SMSQmulator/), [uqlx](http://www.dilwyn.me.uk/emu/index.html#uQLx_for_Linux_etc.), and [Q-emulator](http://www.terdina.net/ql/q-emulator.html).

## Installation

The file QBOXRUN.ZIP is a ready-to-run installation for use with QPC2, SMSQmulator or UQLX. It contains a QBOXRUN.WIN file which contains the QBOX system files, example files and the complete archive of QL file areas of the original QBOX BBS, maintained by Jan Bredenbeek from 1987 to 2003. You should configure your emulator to make device win8_ point to this container file. If you want to use a different device, you have to run QSETUP and change the directories manually. Also, you have to edit the files MAREAS_BBS and FAREAS_BBS which contain the names of the directories where the MESSAGES_BBS files (for MAREAS_BBS) and FILES_BBS reside. You can also use QSETUP to change any settings to suit your preferences.

You can now do a local 'test run' by entering **EX win8_QBOX;'-L -T -D win8_'** from the command line. A terminal program opens and you will see the login page displayed; now you can log in using the user name 'Sysop' and password 'QBOX' (don't forget to change it if you are going to do anything serious!).

You can edit the files LOGON_BBS, LOGOFF_BBS, BULLETIN_BBS, INFOxx_BBS to suit. In order to create new message or file areas, you have to edit the text files MAREAS_BBS and FAREAS_BBS. Each line contains a message or file area definition, which must start with the device and directory name containing the messages or files, followed by a space and the description (see the documentation files for more info). In a message area directory, there should be a file named MESSAGES_BBS (when creating an area you should copy the file EMPTY_MESSAGES_BBS to this directory). In a file area directory, there should be a file FILES_BBS which is a text file containing file names and descriptions of files within the area.

Finally, to start the online BBS, enter **EX win8_QBOX;'-D win8_'** from the command line. If you do **PROG_USE win8_** first then you can simply enter **EX QBOX** as QBOX searches the Toolkit II default program and data directories for its SETUP_BBS file. If all is well then QBOX will start up and print a message 'Waiting for call'. You can now try to connect by starting a Telnet session to the IP address of the machine where QBOX is running, on port 5000 (default setting on QSPILTCP). If QBOX is running on the same machine, use **localhost** or **127.0.0.1** to connect.

## Use in a TCP/IP environment

QBOX was originally written for use with analogue modems, but designed in such a way that the low-level routines which communicate with the 'outside world' were isolated in a so-called QSPIL driver (QSPIL stands for QBOX Serial Port Interface Layer, which is a bit akin to the FOSSIL interface used by DOS-based BBS software at the time). The reason for this was the variety of modems originally in use before most modem manufacturers settled for the 'Hayes' (AT-command) standard. From 1990 onwards, these routines were incorporated in the QSPHAYES driver which is a configurable driver for Hayes-compatible modems.

Now that analogue modems are obsoleted by IP networks and popular QL emulators support communication via the TCP device, the QSPIL interface comes in handy as interface layer between QBOX and IP networks. And so, QSPHAYES has been replaced by QSPILTCP! There is no need to emulate a Hayes-compatible modem as QSPIL can interface to the TCP device directly. Some functions have limitations though depending on the emulator used: 

* Versions of SMSQ/E up to and including 3.35 (used by QPC2 and SMSQmulator) do not signal a TCP connection closed by the client back to QBOX, leading to a hung session when a user closes the connection without properly logging off. Eventually, QBOX will force a user log-off after a timeout of about 3 minutes. In SMSQ/E 3.36 and later, this bug will have been fixed.

* UQLX does signal a closed TCP connection back to QBOX (giving EOF on input), but has a quirk when the connection was closed by QBOX itself after a user log-off: while the closed TCP connection is in a TIME_WAIT state (usually for one minute after closing), uqlx doesn't allow re-binding of a socket to wait for a new connection. QSPILTCP will detect this and pause for one minute before trying again (giving a log message). During this period, it cannot accept new connections.

* On UQLX, it is not possible for QSPILTCP to listen on ports below 1024 since these are 'privileged' ports which can only be bound to by processes running as root. (You can of course circumvent this by running UQLX as root, but this is not desirable from a security point of view). For this reason, QSPILTCP will by default listen on port 5000. If you really need to listen on a privileged port, use iptables to redirect this port to the port QSPILTCP listens on (Example: sudo iptables -t nat -A PREROUTING -p tcp --dport 23 -j DNAT --to-destination :5000)

* Note that UQLX releases earlier than 2018-10-28 have a bug in the TCP driver (it was incorrectly named \*tcp), causing QSPILTCP to fail. The corrected release can be downloaded [here](http://www.dilwyn.me.uk/emu/uqlx2018a.zip).

* For Q-emulator, you need at least version 3.3.1 to run QBOX over TCP/IP successfully.

When an incoming connection has been established, QSPILTCP monitors the first bytes of the incoming data stream for Telnet IAC characters ($FF). If detected, it will send the Telnet options Binary Transmission (WILL/DO), Echo (WILL), and Suppress Go-Ahead (WILL/DO) to initialise the client. This may be changed by editing the teln_opt string. It will also respect the IAC character during send and receive, meaning that a $FF character in the actual data will be escaped by doubling it in the data stream. This will allow file up- and download to work correctly over Telnet sessions when using programs such as TeraTerm or a TCP-to-serial converter such as TCPSER.

If QSPILTCP does not detect the Telnet IAC character during the handshake phase, it will revert to 'raw' mode and send and receive the data 'as-is'. This is fine for terminal programs like QLTERM which can communicate with a Telnet host using redirected I/O but do not support the Telnet IAC escape character by itself. NOTE: you should NOT use the Telnet S\*BASIC procedure in the QLTERM package to start Telnet since this does send a Telnet handshake, which caused QSPILTCP to enable IAC escape which QLTERM doesn't support! Just open a channel using chan=FOP_IN("tcp_A.B.C.D:5000"):EW QLTERM,#chan (of course you should fill in the correct IP address and port number!)

If you want to download files using SEAlink, you might consider turning off the Overdrive option in QSETUP, else it might fail because the large buffers used by TCP connections will cause it to send the whole file at once, causing it to time-out waiting for the acknowledge from the receiver.

Since QBOX was designed to use a modem, it can only handle one connection at a time - subsequent attempts to connect will be refused. However it is still possible to have a remote user *and* a local sysop logged in at the same time. Chat sessions are also supported.

## Configuring QSPILTCP

Although QSPILTCP comes pre-configured with options which I believe are reasonable, it does allow for some configuring. This can either be done by re-assembling the source code (please use the free Quanta/GST assembler followed by the linker to build the binary, do NOT just assemble the _asm with -NOLINK option), or patching the configuration data at the beginning of the QSPILTCP binary using a binary file editor or (B)PUT commands in S\*BASIC. The latter should of course be done with care or you'll end up with an unusable QSPIL. The QSETUP program hasn't been updated so far to include a more user-friendly way, but might be somewhere in the future.

The options which are configurable are:


| Offset | Type | Property                                                                  |
|:------:| ---- |:------------------------------------------------------------------------- |
|   72   | word | Check value, should be 301 decimal                                        |
|   84   | word | TCP port to listen on (default 5000)                                      |
|   86   | long | IP address to listen on (default 0.0.0.0 for 'all')                       |
|   98   | byte | Data transfer protocol to use: 0 raw TCP, 1 Telnet, -1 adaptive (default) |
|  100   |string| Telnet commands to send on connect, only used when above byte is nonzero. Length word followed by max. 30 bytes|
|  132   |string| Name of TCP device (length word followed by max. 30 bytes). Default 'TCP_'|
|  164   |string| Name of SCK device (length word followed by max. 30 bytes). Default 'SCK_'|

## Securing your system

First, a word of warning: QBOX was designed in the 1980s, when security standards were quite different from what we would consider today as minimal. Communication is not encrypted and even the user's password is stored in the USERS_BBS database as plain text! I would recommend *not* using it for communication which should remain private, and even not exposing its IP address to the world. You have been warned!

If you still want to run a publicly available BBS, consider using SSH (Secure Shell) as an extra layer of security. On Linux systems, an SSH server is usually already installed by default. It can be easily configured as a gateway for QBOX using the following steps:

1. Create a new user which will be used as login for the gateway to QBOX. A good example might be 'qbox' (note that, as usual in Linux systems, the name is case sensitive). Assign a password that you can communicate to users (it doesn't have to be secret since users will authenticate themself within QBOX after logging in to SSH). Do *not* use this account for any other purposes; even running the QBOX environment under UQLX should be done under a different user account.

2. Next, edit the /etc/ssh/sshd_config file and include the following lines:

```
        Match User qbox
          DisableForwarding yes
          PermitTTY yes
          ForceCommand telnet -E -8 localhost 5000
```

This locks down the user qbox so after the SSH login it forces execution of the telnet command to connect to the BBS running on the same machine on port 5000. Of course, if your BBS is running on a different port or even a different machine, you should replace 'localhost 5000' to suit. The -E switch avoids the user being able to use the Telnet escape character to escape into Telnet command mode.

## Further reading

If you want to use a QL or compatible system as a BBS client using TCP/IP, please read the [QL-client.md](https://github.com/janbredenbeek/QBOX/blob/master/QL-client.md) file for more info.

Jan Bredenbeek, February 2020.


fidonet/CHANGES.TXT
List of changes for version 1.19j, released 25th March 1994:
------------------------------------------------------------

- Extended the number of QSPIL baudrates supported to allow for sixteen
  possible rates, as described in the new QSPHAYES documentation.
  The new rates supported are 7200, 12000, 14400, 16800, 19200, 21600,
  24000, 26400, 28800 and 38400 bps.


List of changes for version 1.19i, released 26th March 1993:
------------------------------------------------------------

- Fixed some obscure bugs in the SEAlink sender code, which could cause
  SEAlink transmit sessions to fail badly with certain mailers.

- The "product version numbers" in the .PKT header are now correctly set.

- The last block of a sent file is now padded with CTRL-Z's to avoid "junk"
  being added to the file when using XMODEM.

- Not a bugfix, but please note that my FidoNet address has changed from
  2:283/500 to 2:283/1!

   
List of changes for version 1.19h, released 25th December 1992:
---------------------------------------------------------------

- When initiating a plain XMODEM receive, a file length of 65535 was reported.
  Fixed now.

- When receiving, SEAlink Overdrive is now only requested when the inbound 
  device is a RAMdisk. This prevents flow control problems which showed up when 
  receiving using SLO.

- When the receiver switched back to Checksum mode, the SENDNAK routine still
  polled the other end with a C rather than NAK. Fixed now.

  
List of changes for version 1.19g, released 12th April 1992:
------------------------------------------------------------

- Once again made some changes to SEAlink error recovery (comments please,
  Marco!)


List of changes for version 1.19f, released 5th April 1992:
-----------------------------------------------------------

- Improved SEAlink error recovery logic; hopefully it works now...


List of changes for version 1.19e, released 16th February 1992:
---------------------------------------------------------------

- When resuming an interrupted transfer using SEAlink RESYNC, the file date was
  changed after the transfer, so further RESYNCs on the same file were
  impossible (at least on V2 drivers). Fixed now.

- Minor fix: a NAK is now sent when an out-of-sync block has been received.
  Previous versions caused a timeout to occur if this happened.

- SEAlink Overdrive (i.e. "ackless" SEAlink, used on MNP-connections) didn't
  acknowledge EOF. Was already fixed in 1.19d, but I forgot to include it in
  the list...
  

List of changes for version 1.19d, released 11th January 1992:
--------------------------------------------------------------

- Introduced some delay when starting up a session. This will hopefully
  eliminate "fall-through" to the BBS which was reported on earlier versions.

- Many users of Miracle Systems harddisks reported problems, mainly caused by
  the delayed directory update when a new file is created. The SEAlink code now
  incorporates a kludge which hopefully eliminates this; I'm unable to test it
  myself since I don't have a harddisk anymore...

- The SEAlink receive buffer is now also written out if the receive session
  fails for some reason.

- The time stamp stored at offset 10 in the file header, which was removed in
  1.19b, is now back again. The actual problem was that I forgot to clear the
  other header fields before writing out the header...

  
List of changes for version 1.19c, released 27th December 1991:
---------------------------------------------------------------

- Fixed MACFLOW problems which occurred occasionally at end of file;
- Inbound files now don't have junk appended at their end when using Telink or
  SEAlink and the inbound is on RAMdisk.


List of changes for version 1.19b, released 26th December 1991:
---------------------------------------------------------------

- MAIL used a 32K buffer in SEAlink receive when using a winchester disk for
  inbound. Unfortunately, flushing of the entire 32K buffer to disk failed
  because IO.SSTRG fails when the "length" parameter exceeds 32767! The buffer
  size when using win_ has now been reduced by 128 bytes so it will hopefully
  work OK now.

- The time stamp stored at offset 10 decimal in the file header on non-V2
  drivers caused major problems with TOSSMAIL v1.19b, which treated it as an
  offset in the packet file to resume tossing from after an interrupted toss!
  This feature has now temporarily been disabled until a new version of
  TOSSMAIL comes out.

List of changes for version 1.19a, releaased 24th December 1991:
----------------------------------------------------------------

- First new release since v1.04.
  


qsetup/qsetup.c
#include "qsetup_h"

/* #define skipspc(p) while(isspace(*p)) p++; */

char _prog_name[] = "QSETUP";

void (*_consetup)() = NULL;
char *_endmsg = NULL;

struct _setup setup;

char path[17] = "\0";

struct simple_item i_main[] = {
   1, 1, "   Filenames   ",
   1, 2, "Network Mailer ",
   1, 3, "     Users     ",
   1, 4, " Miscellaneous ",
   1, 5, "     Quit      "};

void error(char *msg)
{
   int ww;
   struct _wdef *w;
   char c;

   ww = (strlen(msg) + 2) * 6 + 4;
   if (ww < 172) ww = 172;
   w = w_open(ww, 62, (512 - ww) / 2, 104, W_ETRAP);
   w->paper = w->strip = 210;
   w->bcolour = 2;
   w_setup(w, 2, 2, 7, "ERROR");
   sd_pos(w->chid, -1, 1, 1);
   fputs(msg, w->fp);
   sd_pos(w->chid, -1, 1, 3);
   fputs("Press any key to continue ", w->fp);
   sd_cure(w->chid, -1);
   io_fbyte(w->chid, -1, &c);
   sd_curs(w->chid, -1);
   w_close(w);
}

int getsetup(void)
{
   char sfile[50], temp[50];
   char c;
   char *p;
   int i, fd, extra;
   struct _wdef *w;

   if (!*path) {
      p = getenv("QBOX");
      if (p)
         strcpy(path, p);
      else
         getcwd(path, 24);
   }
   if (!isdevice(path, &extra)) {
      *path = '\0';
      w = w_open(286, 32, 113, 104, W_ETRAP);
      w->paper = w->strip = 210;
      w->estrip = 0;
      w_setup(w, 0, 0, 0, NULL);
      sd_pos(w->chid, -1, 1, 1);
      fputs("Enter QBOX system directory: ", w->fp);
      while (1) {
         if ((i = w_edstr(w, 30, 1, 16, path)) < 0 || i == 0x1B) return -1;
         if (isdevice(path, &extra)) break;
         sd_pos(w->chid, -1, 1, 0);
         sd_clrln(w->chid, -1);
         fprintf(w->fp, "'%s' is not a valid directory", path);
      }
      w_close(w);
   }
   strcpy(sfile, path);
   strcat(sfile, "SETUP_BBS");
   if ((fd = open(sfile, O_RDONLY)) < 0) {
      if (_oserr != ERR_NF) {
         error("Unable to open setup file!");
         return -1;
      }
      w = w_open(280, 32, 116, 104, W_ETRAP);
      w->paper = w->strip = 2;
      w_setup(w, 0, 0, 0, NULL);
      sd_pos(w->chid, -1, 1, 1);
      fputs("Setup file not found, create it? (Y/N/Esc) ", w->fp);
      sd_cure(w->chid, -1);
      do {
         io_fbyte(w->chid, -1, &c);
         c = tolower(c);
      } while (c != 0x1B && c != 'y' && c != 'n');
      sd_curs(w->chid, -1);
      w_close(w);
      if (c != 'y') return -1;
      if ((fd = creat(sfile, 0)) < 0) {
         error("Unable to create setup file!");
         return -1;
      }
      memset(&setup, 0, sizeof(setup));
      setup.fingerprint = SET_FPRINT;
      setup.thisrev = THISREV;
      setup.comprev = COMPREV;
      setup.totlen = sizeof(setup);
      cstr_to_ql((struct QLSTR *)setup.sysdir, path);
      strcpy(temp, path);
      strcat(temp, "IN_");
      cstr_to_ql((struct QLSTR *)setup.inbound, temp);
      strcpy(temp, path);
      strcat(temp, "OUT_");
      cstr_to_ql((struct QLSTR *)setup.outbound, temp);
      cstr_to_ql((struct QLSTR *)setup.msgtmp, "ram1_");
      cstr_to_ql((struct QLSTR *)setup.qspil, "QSPHAYES");
      cstr_to_ql((struct QLSTR *)setup.log1, "ram1_qbox_log");
      cstr_to_ql((struct QLSTR *)setup.log2, "ram1_qbox2_log");
      cstr_to_ql((struct QLSTR *)setup.editor, "flp1_QED");
      cstr_to_ql((struct QLSTR *)setup.terminal, "flp1_QLTERM");
      cstr_to_ql((struct QLSTR *)setup.sysop, "Sysop");
      setup.init_sec = DISGRACE;
      setup.init_mgroups = 0;
      setup.init_fgroups = 0;
      setup.logintime[0] = 10;
      setup.logintime[1] = 20;
      setup.logintime[2] = 30;
      for (i = 3; i < 10; i++) setup.logintime[i] = 60;
      setup.sysopmsg = 0;
      write(fd, &setup, sizeof(setup));
      close(fd);
   }
   else {
      if (read(fd, &setup, sizeof(setup)) < 0) {
         error("Error reading setup file!");
         close(fd);
         return -1;
      }
      close(fd);
      if (setup.fingerprint != SET_FPRINT) {
         error("Setup file is corrupted!");
         return -1;
      }
      if (setup.comprev != COMPREV) {
         error("Setup file version is incompatible!");
         return -1;
      }
   }
   return 0;
}

char savechanges(void)
{
   struct _wdef *w;
   char ret;

   w = w_open(166, 30, 173, 104, W_ETRAP);
   w->paper = w->strip = 2;
   w_setup(w, 0, 0, 0, NULL);
   sd_pos(w->chid, -1, 1, 1);
   fputs("Save changes? (Y/N/Esc) ", w->fp);
   sd_cure(w->chid, -1);
   do {
      io_fbyte(w->chid, -1, &ret);
      ret = tolower(ret);
   } while (ret != 'y' && ret != 'n' && ret != 0x1B);
   sd_curs(w->chid, -1);
   w_close(w);
   return ret;
}

void putsetup(void)
{
   int fd;
   char name[50], oldname[50];

   strcpy(name, path);
   strcat(name, "SETUP_BBS");
   strcpy(oldname, name);
   strcat(oldname, "_OLD");
   unlink(oldname);
   rename(name, oldname);
   if ((fd = creat(name, 0)) < 0) {
      error("Unable to save setup file!");
      return;
   }
   if (write(fd, &setup, sizeof(setup)) != sizeof(setup)) {
      error("Writing new setup file failed (disk full?)");
      close(fd);
      unlink(name);
      return;
   }
   close(fd);
}

char *fnam_opt[] = {"    System directory:",
                    "   Inbound directory:",
                    "  Outbound directory:",
                    " Temp. Msg directory:",
                    "   Remote server log:",
                    "    Local server log:",
                    "Local message editor:",
                    "   Terminal emulator:",
                    "   QSPIL driver name:"};

char *fnam_hlp[] = {"Where the system files are stored",
                    "Where incoming network files are stored",
                    "Where outgoing network files are stored",
                    "Used for external message editor workfile",
                    "Logfile for remote logons",
                    "Logfile for local logons",
                    "Full filename of text editor for local messages",
                    "Full filename of terminal emulator (local & chat)",
                    "Filename (NOT directory) of QSPIL driver"};

void act_filenames(struct _wdef *w, int item, int flag)
{
   struct QLSTR *qs;
   char temp[50];
   int maxlen, extra;

   if (flag > 0) {
      sd_pos(w->chid, -1, 1, 11);
      sd_setpa(w->chid, -1, w->estrip);
      sd_setst(w->chid, -1, w->estrip);
      sd_setin(w->chid, -1, w->eink);
      sd_clrln(w->chid, -1);
      fputs(fnam_hlp[item], w->fp);
   }
   sd_pos(w->chid, -1, 1, 1 + item);
   sd_setst(w->chid, -1, (flag > 0) ? w->sstrip : w->strip);
   sd_setin(w->chid, -1, (flag > 0) ? w->sink : w->ink);
   fputs(fnam_opt[item], w->fp);
   sd_setst(w->chid, -1, w->strip);
   sd_setin(w->chid, -1, w->ink);
   fputc(' ', w->fp);
   switch (item) {
   case 0: qs = (struct QLSTR *) setup.sysdir;
           maxlen = MAXPATH - 2;
           break;
   case 1: qs = (struct QLSTR *) setup.inbound;
           maxlen = MAXPATH - 2;
           break;
   case 2: qs = (struct QLSTR *) setup.outbound;
           maxlen = MAXPATH - 2;
           break;
   case 3: qs = (struct QLSTR *) setup.msgtmp;
           maxlen = MAXFNAME - 2;
           break;
   case 4: qs = (struct QLSTR *) setup.log1;
           maxlen = MAXFNAME - 2;
           break;
   case 5: qs = (struct QLSTR *) setup.log2;
           maxlen = MAXFNAME - 2;
           break;
   case 6: qs = (struct QLSTR *) setup.editor;
           maxlen = MAXFNAME - 2;
           break;
   case 7: qs = (struct QLSTR *) setup.terminal;
           maxlen = MAXFNAME - 2;
           break;
   case 8: qs = (struct QLSTR *) setup.qspil;
           maxlen = 20;
           break;
   }
   if (flag == I_ACTION) {
      qlstr_to_c(temp, qs);
      w_edstr(w, 23, item + 1, maxlen, temp);
      if (item != 8 && !isdevice(temp, &extra)) {
         sd_pos(w->chid, -1, 1, 11);
         sd_setpa(w->chid, -1, w->sstrip);
         sd_setst(w->chid, -1, w->sstrip);
         sd_setin(w->chid, -1, w->sink);
         sd_clrln(w->chid, -1);
         fputs("Unable to locate directory!", w->fp);
      }
      cstr_to_ql(qs, temp);
   }
   else io_sstrg(w->chid, -1, qs->qs_str, qs->qs_strlen);
}


void do_filenames(void)
{
   struct _wdef *w;

   w = w_open(412, 132, 50, 64, W_ETRAP);
   w->paper = w->strip = 0;
   w->action = act_filenames;
   w->n_items = w->w_items = 9;
   w_setup(w, 116, 2, 7, " Filenames ");
   w_select(w, NULL, "\x1B");
   w_close(w);
}

void time_to_asc(char *buf, unsigned short *val)
{
   sprintf(buf, "%2d:%02d", *val / 60, *val % 60);
}

int asc_to_time(unsigned short *ret, char *buf)
{
   int h, m;

   if (strchr(buf, ':')) {
      if (sscanf(buf, "%d:%d", &h, &m) != 2) return -1;
   }
   else {
      m = atoi(buf);
      h = m / 100; m %= 100;
   }
   if (h < 0 || h > 23 || m < 0 || m > 59) return -1;
   *ret = h * 60 + m;
   return 0;
}

void addr_to_asc(char *buf, ADDRESS *addr)
{
   if (!addr->net) {
      *buf = '\0';
      return;
   }
   sprintf(buf, "%u:%u/%u.%u", addr->zone, addr->net, addr->node, addr->point);
}

int readnum(char **pp)
{
   int ret;
   char *p;

   p = *pp;
   ret = (int) strtoul(p, pp, 10);
   return (p == *pp) ? -1 : ret;
}

/* Read network address (may be incomplete)
   Returns 0 if OK, -1 if syntax error */

int asc_to_addr(ADDRESS *addr, char *p)
{
   int ret;

   skipspc(p);
   if (!*p) {
      memset(addr, 0, sizeof(ADDRESS));
      return 0;
   }
   if (*p == '.') goto readpoint; /* short point form */

   if ((ret = readnum(&p)) < 0) return ret;
   if (*p == ':') {  /* number was zone number */
      p++;
      addr->zone = ret;
      if ((ret = readnum(&p)) < 0) return ret;
      if (*p != '/') return -1;
   }
   if (*p == '/') {  /* number was net number */
      p++;
      addr->net = ret;
      if ((ret = readnum(&p)) < 0) return ret;
   }
   addr->node = ret;
   if (*p != '.') return 0;  /* Done if no point specified */

readpoint:
   p++;
   if ((ret = readnum(&p)) < 0) return ret;
   addr->point = ret;
   return 0;
}

void pn_to_asc(char *buf, short int *pn)
{
   *buf = '\0';
   if (*pn) sprintf(buf, "%d", *pn);
}

int asc_to_pn(short int *pn, char *buf)
{
   *pn = atoi(buf);
   if (*pn <= 0) *pn = 0;
   return 0;
}

char *mailer_opt[] = {"           Addresses ",
                      "Mail-only start time:",
                      "  Mail-only end time:"};

char *addr_opt[] = {"  Main:",
                    "AKA #1:",
                    "AKA #2:",
                    "AKA #3:",
                    "AKA #4:",
                    "AKA #5:",
                    "AKA #6:",
                    "AKA #7:",
                    "AKA #8:",
                    "AKA #9:"};

void act_addr(struct _wdef *w, int item, int flag)
{
   char buf[25];

   sd_pos(w->chid, -1, 1, 2 + item);
   sd_setst(w->chid, -1, (flag > 0) ? w->sstrip : w->strip);
   sd_setin(w->chid, -1, (flag > 0) ? w->sink : w->ink);
   fputs(addr_opt[item], w->fp);
   sd_setst(w->chid, -1, w->strip);
   sd_setin(w->chid, -1, w->ink);
   fputc(' ', w->fp);
   if (flag == I_ACTION) {
      w_edit(w, 9, 2 + item, 23, &setup.myaddress[item], addr_to_asc,
                asc_to_addr);
      w_edit(w, 33, 2 + item, 5, &setup.pointnet[item], pn_to_asc, asc_to_pn);
   }
   else {
      addr_to_asc(buf, &setup.myaddress[item]);
      fputs(buf, w->fp);
      pn_to_asc(buf, &setup.pointnet[item]);
      sd_tab(w->chid, -1, 33);
      fputs(buf, w->fp);
   }
}

void do_addr(void)
{
   struct _wdef *w;

   w = w_open(238, 142, 136, 54, W_ETRAP);
   w->paper = w->strip = 0;
   w->action = act_addr;
   w->n_items = w->w_items = 10;
   w_setup(w, 116, 2, 7, " Addresses ");
   sd_pixp(w->chid, -1, 6, 5);
   sd_setin(w->chid, -1, 4);
   sd_setul(w->chid, -1, 1);
   fputs("Address                     Pointnet", w->fp);
   sd_setul(w->chid, -1, 0);
   w_select(w, NULL, "\x1B");
   w_close(w);
}

void act_mailer(struct _wdef *w, int item, int flag)
{
   char buf[25];

   sd_pos(w->chid, -1, 1, 1 + item);
   sd_setst(w->chid, -1, (flag > 0) ? w->sstrip : w->strip);
   sd_setin(w->chid, -1, (flag > 0) ? w->sink : w->ink);
   fputs(mailer_opt[item], w->fp);
   sd_setst(w->chid, -1, w->strip);
   sd_setin(w->chid, -1, w->ink);
   fputc(' ', w->fp);
   if (item == 0) {
      if (flag == I_ACTION) do_addr();
   }
   else {
      if (flag == I_ACTION)
         w_edit(w, 23, 1 + item, 5, (item == 1) ? &setup.mail_start :
                &setup.mail_end, time_to_asc, asc_to_time);
      else {
         time_to_asc(buf, (item == 1) ? &setup.mail_start : &setup.mail_end);
         fputs(buf, w->fp);
      }
   }
}

void do_mailer(void)
{
   struct _wdef *w;

   w = w_open(178, 62, 166, 94, W_ETRAP);
   w->paper = w->strip = 0;
   w->action = act_mailer;
   w->n_items = w->w_items = 3;
   w_setup(w, 116, 2, 7, " Mailer ");
   w_select(w, NULL, "\x1B");
   w_close(w);
}

char *user_lvl[] = {"Twit", "Disgrace", "Normal", "Special", "Extra",
                    "", "", "", "CoSysOp", "SysOp"};

char *user_opt[] = {"       New user security level:",
                    "  New user message area groups:",
                    "     New user file area groups:",
                    "  Minutes per day for '  Twit':",
                    "Minutes per day for 'Disgrace':",
                    "  Minutes per day for 'Normal':",
                    " Minutes per day for 'Special':",
                    "   Minutes per day for 'Extra':",
                    " Minutes per day for 'CoSysOp':",
                    "   Minutes per day for 'SysOp':"};

void u_ttoa(char *buf, short *val)
{
   sprintf(buf, "%d", (int) *val);
}

int u_atot(short *val, char *buf)
{
   int i;

   i = atoi(buf);
   if (i < 0 || i > 1440) return -1;
   *val = i;
   return 0;
}

void u_gtoa(char *buf, unsigned long *val)
{
   char c = 'A';
   unsigned long mask = 1;

   while (c <= 'Z') {
      if (*val & mask) *buf++ = c;
      c++;
      mask <<= 1;
   }
   *buf = '\0';
}

int u_atog(unsigned long *val, char *buf)
{
   char c;
   unsigned long result = 0L;

   while ((c = *buf++) != '\0') {
      c = toupper(c);
      if (c < 'A' || c > 'Z') return -1;
      result |= 1 << (c - 'A');
   }
   *val = result;
   return 0;
}

void act_user(struct _wdef *w, int item, int flag)
{
   char buf[27];

   sd_pos(w->chid, -1, 1, 1 + item);
   sd_setst(w->chid, -1, (flag > 0) ? w->sstrip : w->strip);
   sd_setin(w->chid, -1, (flag > 0) ? w->sink : w->ink);
   fputs(user_opt[item], w->fp);
   sd_setst(w->chid, -1, w->strip);
   sd_setin(w->chid, -1, w->ink);
   fputc(' ', w->fp);
   switch (item) {
   case 0:
      if (flag == I_ACTION) {
         setup.init_sec++;
         if (setup.init_sec > MAXLEV) setup.init_sec = 0;
         while (!*user_lvl[setup.init_sec]) setup.init_sec++;
         sd_clrrt(w->chid, -1);
         fputs(user_lvl[setup.init_sec], w->fp);
      }
      else fputs(user_lvl[setup.init_sec], w->fp);
      break;
   case 1:
   case 2:
      if (flag == I_ACTION)
         w_edit(w, 33, 1 + item, 26, (item == 1) ? &setup.init_mgroups :
                &setup.init_fgroups, u_gtoa, u_atog);
      else {
         u_gtoa(buf, (item == 1) ? &setup.init_mgroups : &setup.init_fgroups);
         fputs(buf, w->fp);
      }
      break;
   case 3: case 4: case 5: case 6: case 7: case 8: case 9:
      if (flag == I_ACTION)
         w_edit(w, 33, 1 + item, 4, &setup.logintime[item - 3], u_ttoa, u_atot);
      else
         fprintf(w->fp, "%d", setup.logintime[item - 3]);
      break;
   }
}

void do_user(void)
{
   struct _wdef *w;

   w = w_open(364, 132, 74, 64, W_ETRAP);
   w->paper = w->strip = 0;
   w->action = act_user;
   w->n_items = w->w_items = 10;
   w_setup(w, 116, 2, 7, " Users ");
   w_select(w, NULL, "\x1B");
   w_close(w);
}

char *misc_opt[] = {"                       System name:",
                    "                        SysOp name:",
                    "                   Local mode only:",
                    "                         Mail only:",
                    "                      'Yell' alarm:",
                    "                        Start time:",
                    "                          End time:",
                    "Message area for messages to SysOp:"};

char *misc_hlp[] = {"Name of your Bulletin Board System",
                    "Name of System Operator",
                    "If 'YES', QBOX starts up as if '-L' switch was given",
                    "If 'YES', QBOX only accepts Network Mail calls",
                    "If 'YES', QBOX sounds alarm during [Y]ell command",
                    "Start time for Yell alarm if enabled",
                    "End time for Yell alarm if enabled",
                    "Number of area for leaving comments etc. (0 = disabled)"};

int main(int argc, char *argv[], char *envp[])
{
   struct _wdef *w, *main_w;
   int i;
   char c;

   w_paper = w_strip = 0;
   w_ink = 7;
   w_sstrip = 4;
   w_sink = 0;
   w_estrip = 2;
   w_eink = 7;
   w_bcolour = 4;
   w_bwidth = 1;

   if (argc > 1)
      strcpy(path,argv[1]);
   w = w_make(stdout, 448, 182, 32, 24, W_ETRAP);
   w->paper = w->strip = 82;
   w->ink = 7;
   w->bcolour = 7;
   w->bwidth = 1;
   w_setup(w, 111, 7, 0, " QBOX SETUP ");
   sd_pos(w->chid, -1, 28, 1);
   printf("QBOX SETUP v%s", version);
   sd_pos(w->chid, -1, 19, 3);
   printf("Copyright (C) 1992 by Jan Bredenbeek");
   if (getsetup() != 0) goto exit_setup;
   main_w = w_open(106, 82, 202, 94, W_ETRAP | W_NOSAV);
   main_w->ilist = i_main;
   main_w->n_items = main_w->w_items = 5;
   while (1) {
      w_save(main_w);
      w_setup(main_w, 116, 2, 7, " Main ");
      i = w_select(main_w, "fnumq", "\n\x1B fnumq");
      w_restore(main_w);
      if (i == 0x1B || main_w->c_item == 4) {
         c = savechanges();
         if (c == 0x1B) continue;
         if (c == 'y') putsetup();
         break;
      }
      switch(main_w->c_item) {
      case 0: do_filenames(); break;
      case 1: do_mailer(); break;
      case 2: do_user(); break;
/*      case 3: do_misc(); break;     */
      }
   }

   w_close(main_w);
exit_setup:
   w_close(w);
   return 0;
}

